### PR TITLE
Thermomachine rework

### PR DESCRIPTION
**Pull request by Ghilker**
## About The Pull Request
Freezers/heaters now works as heat pumps, that moves heat from one source to another, there are now 2 ports on the machine, one green (that will work as the current one, where you connect the pipes and cool down/heat up the contents) and an optional red port that allows the heat movement.
IF no pipe is connected to the red port, or it doesn't have any gases in it, the thermomachine will use the air around it to move the heat around, cooling or heating up the air depending on the settings
Update: Freezers in map are now connected to the waste loop and the CO2 is getting filtered back into waste to be used as heat collector too, atmos now should manage this loop more actively.
TO DO
- [x] Better checks for heat movement
- [x] Maybe heaters don't need the red port connected (they could use much more power to work tho, gimme your thoughts)

## Why It's Good For The Game
freezers now don't delete heat
This change will help move forward to heat based engine and a better experience for the heat management systems.

## Changelog
:cl:
balance: READ ME thermomachines now have 2 ports, the green one is used as you normally do, by cooling/heating the gases connected to it, the red port is optional but it helps move the heat from the green port to the red one (and vice-versa), if no pipe is connected to the red port or there is no gas in it, the heat will be moved from the enviroment; 
balance: ALL the station freezers are now connected to the waste loop, where the CO2 gets collected and used to store that heat (atmos should now manage the waste pipenet more actively); 
balance: the best way to now use freezers is to connect their red port to the waste system (there is at least one pipe in every room)
balance: Rotate the machine without emptying the content will release the gas onto the atmosphere
/:cl:

### CODE REVIEW

**Comment by zxaber** [comment]
Heaters could allow some heat generation on their own, with both a much higher power cost and also a cap on max temps.

How do you have rotations done? Currently a wrench will rotate for gas pipe connections, but with two connections that gets fuzzy. Do both ports always point the same way?

This is probably going to affect toxins a lot, are they still able to make their bombs with round-start setups?

**Comment by Ghilker** [comment]
> Heaters could allow some heat generation on their own, with both a much higher power cost and also a cap on max temps.

I was thinking about this yes, heater can use just power instead of moving heat around 

> How do you have rotations done? Currently a wrench will rotate for gas pipe connections, but with two connections that gets fuzzy. Do both ports always point the same way?

Rotation will happen normally, with the ports moving around the machine and staying at 180 degrees apart, can't make them rotate individually, current atmosmachinery can't handle it

> This is probably going to affect toxins a lot, are they still able to make their bombs with round-start setups?

Nothing else changed, only that if toxins doesn't hook it up to some cooling source (like space) the room will get hotter, but a space heater is more than enough to contrast it

**Comment by D4C-420** [comment]
How is this going to affect round start xenobio because they rely on a cooler to make the cold room, cold enough to kill slimes. The other concern I have is for RND's server room which also is kept cool by 2 coolers. Some map edits might not be a bad idea depending on how hot the air is heated near the servers and xenobio.

**Comment by interestingusernam3** [comment]
atmos gaming

**Comment by Ghilker** [comment]
> How is this going to affect round start xenobio because they rely on a cooler to make the cold room, cold enough to kill slimes. The other concern I have is for RND's server room which also is kept cool by 2 coolers. Some map edits might not be a bad idea depending on how hot the air is heated near the servers and xenobio.

Once the temperature stabilizes there won't be more issues about heating, when the machine reaches the requested temperature it basically stops

If there should be issues I'll do some map edits but I don't think there will be any

**Comment by SuperNovaa41** [comment]
> > How is this going to affect round start xenobio because they rely on a cooler to make the cold room, cold enough to kill slimes. The other concern I have is for RND's server room which also is kept cool by 2 coolers. Some map edits might not be a bad idea depending on how hot the air is heated near the servers and xenobio.
> 
> Once the temperature stabilizes there won't be more issues about heating, when the machine reaches the requested temperature it basically stops
> 
> If there should be issues I'll do some map edits but I don't think there will be any 

You should test it before the merge so no need to make another PR for map edits ;)

**Comment by zxaber** [comment]
>Nothing else changed, only that if toxins doesn't hook it up to some cooling source (like space) the room will get hotter, but a space heater is more than enough to contrast it

Don't maxcaps use a tank of superchilled oxygen? I admit I don't make a lot of atmos bombs.

**Comment by SuperNovaa41** [comment]
> > Nothing else changed, only that if toxins doesn't hook it up to some cooling source (like space) the room will get hotter, but a space heater is more than enough to contrast it
> 
> Don't maxcaps use a tank of superchilled oxygen? I admit I don't make a lot of atmos bombs.

Yes. Oxygen or trit/oxygen

File: code/modules/atmospherics/machinery/components/binary_devices/thermomachine.dm [hunk 1]
```diff
@@ -117,31 +121,37 @@
...
+	var/datum/gas_mixture/thermal_exchange_port = airs[2]
+	var/turf/local_turf = get_turf(src)
+	var/main_heat_capaticy = main_port.heat_capacity()
+	var/thermal_heat_capacity = thermal_exchange_port.heat_capacity()
+	var/temperature_delta = main_port.temperature - target_temperature
+
+	var/heat_amount = temperature_delta * (main_heat_capaticy * heat_capacity / (main_heat_capaticy + heat_capacity))
+	if(main_port.total_moles() && thermal_exchange_port.total_moles())
+		main_port.temperature = max(main_port.temperature - heat_amount / main_heat_capaticy, TCMB)
+		thermal_exchange_port.temperature = max(thermal_exchange_port.temperature + heat_amount / thermal_heat_capacity, TCMB)
```
Comment (ninjanomnom) [inline code/modules/atmospherics/machinery/components/binary_devices/thermomachine.dm:139]: Add just a little bit of inefficiency and create some additional heat from the machine itself running

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/modules/atmospherics/machinery/components/binary_devices/thermomachine.dm [hunk 2]
```diff
@@ -117,31 +121,37 @@
...
-
-	if(combined_heat_capacity > 0)
-		var/combined_energy = heat_capacity * target_temperature + air_heat_capacity * air_contents.temperature
-		air_contents.temperature = combined_energy/combined_heat_capacity
-
-	var/temperature_delta = abs(old_temperature - air_contents.temperature)
+	var/datum/gas_mixture/main_port = airs[1]
+	var/datum/gas_mixture/thermal_exchange_port = airs[2]
+	var/turf/local_turf = get_turf(src)
+	var/main_heat_capaticy = main_port.heat_capacity()
```
Comment (rdragan) [inline code/modules/atmospherics/machinery/components/binary_devices/thermomachine.dm:132]:
```suggestion
	var/main_heat_capacity = main_port.heat_capacity()
```

Small typo in this var

**Comment by Ghilker** [comment]
> Don't maxcaps use a tank of superchilled oxygen? I admit I don't make a lot of atmos bombs.

yes but you can still make superchilled o2

**Comment by ninjanomnom** [comment]
The code seems good but I'd like to testmerge this for at least a little while due to the widespread nature of the gameplay changes being made.

**Comment by LemonInTheDark** [comment]
How will this effect gameplay? Thermo machines are used in toxins, atmos, and the sm. I'm most worried about the effect on supermatter work, since I'm reasonably sure people will just pop a hole in the floor, which just pigeon holes plasmamen even further.

The same can kind of be said of toxins, though I'm less sure of this.

**Comment by Ghilker** [comment]
> How will this effect gameplay? Thermo machines are used in toxins, atmos, and the sm. I'm most worried about the effect on supermatter work, since I'm reasonably sure people will just pop a hole in the floor, which just pigeon holes plasmamen even further.
> 
> The same can kind of be said of toxins, though I'm less sure of this.

ill add a check for the turf existing under the freezer and a check for environmental moles. Gameplay doesn't change much, you just have to be more careful when working with them (yes a t4 might kill your APC and yourself too)
Or make a space loop for better efficiency
Or use a spess heater to deal with the room heating up

**Comment by zxaber** [comment]
Rather than a turf check, you could just make the heat transfer speed scale with room pressure when in that mode. As a room is vented to space, heat transfer rate approaches zero.

**Comment by Ghilker** [comment]
> Rather than a turf check, you could just make the heat transfer speed scale with room pressure when in that mode. As a room is vented to space, heat transfer rate approaches zero.

uhh clever idea, yes i'll do that

**Comment by Jhyrachy** [comment]
I have... mixed feeling about that PR

I get that the ability to just delete heat is strong, but at the same time this makes get rid of huge amount of heat really difficult: i tried cooling a 44k kelvin canister (half of a burn can maximum temperature, so not even THAT hot) and it made an hole on the ground in few seconds with 'air cooling'

I somehow like the idea that you need 'liquid cooling' but this just make the freezer act as a middle men between a space loop and the heat source. at that point I can just use an heat pump to the space loop and close it when i'm at the right temperature.
Much faster and require less materials (heat pump is in the pipe dispenser) and research.

Also, there is another problem, that is actual available space.  
Would be trivial for me to attach a small spaceloop on a side and the heat source on the other, the problem is that I need to fill the space loop with something otherwise it will not work, and filling it requires the space for at least a pump, and maybe a connector.
Except one map (delta station) all the other atmos are really cramped to the point that adding a new cooling system for the freezers would rob even more space.

Would be cool if freezers could accept canister and dump them in 'waste' so we could fill the other side with the gas we want without the need of more pumps and manifolds

**Comment by Ghilker** [comment]
you CAN just attach canisters filled with plasma to the red port and have that stores the heat, nothing is stopping you there. 

The idea here is that machines will take a lot more energy to move heat from a very high amount (like the 44k K canister you had) to a low amount. Also you made a hole because you didn't connect anything to the red port

**Comment by Rover115** [comment]
Being honest i think this brings a bit more complexity to the temp management side of the gameplay and makes it more realistic, however overall i believe this will have a net negative impact on gameplay as previously by other people Thermomachines are widely present across the station not just to Atmos/Engineering, Even in the Engineering/Atmos departments this merely further complicates and adds time to round setup and more the setup of more complex projects in which time can already be constrained quite heavily. Overall i dont think this change would improve gameplay for the average spassman nor improve it for those messing in the more complex side of these roles. Realism and complexity isnt always better, Sometimes simple is good.

**Comment by Ghilker** [comment]
The roundstart freezers are all connected to the waste loop now, and in there it will be collected all the CO2 that will be produced during the round, this way if you don't want to make a specialized space loop, you can just connect the red port to the waste loop and let atmos techs manage the heat generated on the station.

**Comment by Fikou** [comment]
split the changelog to multiple balance tags, this will be really long and nobody will read it

**Comment by Ghilker** [comment]
updated to 3 balance tags, if people won't read it, don't blame me

**Comment by vincentiusvin** [comment]
The old thermomachine works on a very simple and elegant basis. We have an object with a set heat capacity with a set temperature open which we interact with. This object is then combined with the gas mixture and thus produces temp changes. The power usage can then scale based on this object alone (once we have a proper power system, that is.)  It's beautiful, simple, elegant.
I'm still a little unclear on how the thermomachine is supposed to work in the PR, like- it has three different heat capacity now? It works based off of share? We share the main port with the freezer's heat capacity and then have the output port be influenced by this? If the purpose is to preserve thermodynamics then why have the efficiency variable? I'm not sure people understand the underlying premise of the PR here. The best bit of the atmospherics code are the parts where it makes sense what is being done that you can imagine it happening in real life, albeit with some suspension of disbelief. Can the working principles of the new thermomachine can be simplified/explained further? Maybe ditch the freezer heat capacity variable and make it work like heat exchangers but with temp checks? Maybe remove the efficiency bit? Maybe have the matter bins be a "heat storage" to store the heat and reduce how effective the freezer is when not hooked to a waste line?

Other than that, looks cool.

**Comment by Ghilker** [comment]
the thermomachine now works by calculating a var called heat_amount
`var/heat_amount = temperature_delta * (main_heat_capacity * heat_capacity / (main_heat_capacity + heat_capacity))` 
this var indicates the amount of heat required to move the temperature from the current temperature to the selected one (temperature_delta), there are 2 heat capacity here, the main one (that is that of the gas inside the green connector) and the heat_capacity of the machine (t1 is 5000).

after calculating the heat_amount var, we check if the machine has a connection to the red port and if that connection has gases in it (if it doesn't the heat get moved to the air around the machine)
`thermal_exchange_port.temperature = max(thermal_exchange_port.temperature + heat_amount / thermal_heat_capacity + motor_heat / thermal_heat_capacity, TCMB)` 
so now we change that temperature to what the red port should be by adding the heat_amount divided by the red port heat capacity AND by adding the small motor heat produced by the machine itself.

After this we calculate the efficiency by checking the log scale of the temperature delta between the thermal port and the main port and then we calculate the temperature of the green port 
`main_port.temperature = max(main_port.temperature - (heat_amount * efficiency)/ main_heat_capacity + motor_heat / main_heat_capacity, TCMB)` 
multiplying the heat moved by the efficiency (better efficiency is achieved by constantly cooling the red port)

this is as simplified as i can get the system without removing the features that makes it so

**Comment by Jhyrachy** [comment]
> 
> 
> you CAN just attach canisters filled with plasma to the red port and have that stores the heat, nothing is stopping you there.
> 
> The idea here is that machines will take a lot more energy to move heat from a very high amount (like the 44k K canister you had) to a low amount. Also you made a hole because you didn't connect anything to the red port

I'm sorry, I think i may have expressed me wrong
immagine

This mockup is how I would make the smallest 'cooling unit'
My point about adding plasma was that, right now, we have to do it with the bottom method, aka add 2 more tiles to add the cooling system.

What I was suggesting was to let us insert tank (like the plasma tank top left of the pic) in the freezers/cooler to add/remove gas from the hot side, like a small air pump.

That would allow us to do the top setup, that is much more compact

Still, I see you already mitigated this connecting the heat side to the waste loop.

**Comment by Ghilker** [comment]
@Jhyrachy i like that idea, it should help a lot in the construction of smal systems
this is how the UI and machine looks like when a tank is inserted
image

This is how the UI looks like when the tank is empty
image

This is how the UI looks like when the tank has been removed
image

File: code/modules/atmospherics/machinery/components/binary_devices/thermomachine.dm [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+	. += getpipeimage(icon, "pipe", dir, COLOR_LIME, piping_layer)
+	. += getpipeimage(icon, "pipe", turn(dir, 180), COLOR_MOSTLY_PURE_RED, piping_layer)
+	if(holding)
+		var/mutable_appearance/holding = mutable_appearance(icon, "holding")
+		. += holding
+
+/obj/machinery/atmospherics/components/binary/thermomachine/update_icon_nopipes()
+	cut_overlays()
+	if(showpipe)
+		add_overlay(getpipeimage(icon, "scrub_cap", initialize_directions))
```
Comment (ninjanomnom) [inline code/modules/atmospherics/machinery/components/binary_devices/thermomachine.dm:105]: If you're messsing with overlays dont be clearing them yourself like this, put this in update_overlays as they get called around the same time anyway

File: code/modules/atmospherics/machinery/components/binary_devices/thermomachine.dm [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+
+/obj/machinery/atmospherics/components/binary/thermomachine/process_atmos()
+	if(!is_operational || !on)  //if it has no power or its switched off, dont process atmos
+		on = FALSE
+		update_appearance()
+		return
+
+	var/turf/local_turf = get_turf(src)
+	var/datum/gas_mixture/enviroment = local_turf.return_air()
+	if(!local_turf)
```
Comment (ninjanomnom) [inline code/modules/atmospherics/machinery/components/binary_devices/thermomachine.dm:134]: If this is true, you would have already thrown an exception on the previous line

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/modules/atmospherics/machinery/components/binary_devices/thermomachine.dm (hunk 3 above)
Comment (Ghilker) [reply code/modules/atmospherics/machinery/components/binary_devices/thermomachine.dm]: this was the old way to do this, now i moved it to update_overlays()

File: code/modules/atmospherics/machinery/components/binary_devices/thermomachine.dm (hunk 4 above)
Comment (Ghilker) [reply code/modules/atmospherics/machinery/components/binary_devices/thermomachine.dm:130]: moved above

**Comment by stylemistake** [comment]
Someone is finally doing justice to thermodynamics!

**Review by stylemistake — APPROVED** [review]

**Review by ninjanomnom — APPROVED** [review]

**Comment by rharvest** [comment]
No mention that you removed the functionality of working machines returning to their on state after power loss?
Aside from that, I loathe this change. The half-assedness of dumping heat into the existing waste loop is the eptiome of externalizing harm.

**Comment by Akride** [comment]
Hey man, I think this is a really cool merge, however I have a few things to add.

First of all, most coolers are placed against walls and their red output is usually not plugged to the waste loop.

Second, I have noticed some issues when working on toxins : 
 -TCUs don't recognize waste (and probably all other pipings) as a suitable environment when plugged to it via a layer adapter. This is a real bother, since you need to move freezers around, and they start on layer 3, while waste is on layer 2. (This is actualy probably why freezers don't work at shift start now that I think about it).  If you could look into that that would be grand. (To reproduce, simply plug a TCU to a layer adapter, and have a layer 2 pipe join with waste loop. The TCU should not detect an environment on the red piping.

-I have noticed coolers have trouble reaching the targeted temp. Take the example of tritium. Tritium always stabilizes at ~53K when I set the cooler to the usual 43K. Coolers seem to cool gasses down to (targeted temp + 10)K. This is a major setback for toxins since you need to upgrade the coolers with even better lasers so you can target 33K and fiddle with stuff more. This is not the end of the world but something to look into definitely.

**Comment by vincentiusvin** [comment]
> The old thermomachine works on a very simple and elegant basis. We have an object with a set heat capacity with a set temperature open which we interact with. This object is then combined with the gas mixture and thus produces temp changes. The power usage can then scale based on this object alone (once we have a proper power system, that is.) It's beautiful, simple, elegant.
> 
> I'm still a little unclear on how the thermomachine is supposed to work in the PR, like- it has three different heat capacity now? It works based off of share? We share the main port with the freezer's heat capacity and then have the output port be influenced by this? If the purpose is to preserve thermodynamics then why have the efficiency variable? I'm not sure I understand the underlying premise of the PR here. The best bit of the atmospherics code are the parts where it makes sense what is being done that you can imagine it happening in real life, albeit with some suspension of disbelief. Can the working principles of the new thermomachine be simplified/explained further? Maybe ditch the freezer heat capacity variable and make it work like heat exchangers but with temp checks? Maybe remove the efficiency bit? Maybe have the matter bins be a "heat storage" to store the heat and reduce how effective the freezer is when not hooked to a waste line?
> 
> Other than that, looks cool.

I think i need to correct myself now, for posterity. It turns out that the heat calculation by the thermomachine from this PR is the exact same as the old one when the efficiency value is equal to one (discounting motor heat). I made this comment without knowing this, and with recently acquired hindsight, I feel like I need to say that the math behind the new heat calculation is actually robust.

image
LHS is the old one, RHS is the new one.

Kudos to the author for figuring this out.

**Pull request merged**